src/test/java/uk/gov/ch/oauth/OAuthCoordinatorTest.java

```diff
@@ -0,0 +1,166 @@
+
+    @Mock
+    private HttpServletResponse mockResponse;
+    @Mock
+    private IIdentityProvider mockIdentityProvider;
+    @Mock
+    private IOauth mockOAuth;
+
+    @Nested
+    class CallbackTests {
+
+        @Test
+        @DisplayName("Throws Exception if callback params contains error.")
+        void getPostCallbackRedirectURLWhenErrorResponse() throws IOException {
+            Map<String, String> params = new HashMap<>();
+            params.put("error", "Error Value");
+
+            assertThrows(UnauthorisedException.class,
+                    () -> oAuthCoordinator.getPostCallbackRedirectURL(mockResponse, params));
+
+            verify(mockResponse).sendError(HttpServletResponse.SC_FORBIDDEN, "Error Value");
+        }
+
+        @Test
+        @DisplayName("Throws Exception if callback do not contain State.")
+        void getPostCallbackRedirectURLWhenMissingValuesState()
+                throws UnauthorisedException, IOException {
+            Map<String, String> params = new HashMap<>();
+            params.put("code", "Code");
+
+            assertThrows(UnauthorisedException.class,
+                    () -> oAuthCoordinator.getPostCallbackRedirectURL(mockResponse, params));
+
+            verify(mockResponse)
+                    .sendError(HttpServletResponse.SC_FORBIDDEN, "Invalid access token");
+            verify(oAuthCoordinator, never())
+                    .validateResponse(anyString(), anyString(), any(HttpServletResponse.class));
+        }
+
+        @Test
+        @DisplayName("Throws Exception if callback do not contain Code.")
+        void getPostCallbackRedirectURLWhenMissingValuesCode()
+                throws UnauthorisedException, IOException {
+            Map<String, String> params = new HashMap<>();
+            params.put("state", "State");
+
+            assertThrows(UnauthorisedException.class,
+                    () -> oAuthCoordinator.getPostCallbackRedirectURL(mockResponse, params));
+
+            verify(mockResponse)
+                    .sendError(HttpServletResponse.SC_FORBIDDEN, "Invalid access token");
+            verify(oAuthCoordinator, never())
+                    .validateResponse(anyString(), anyString(), any(HttpServletResponse.class));
+        }
+
+        @Test
+        @DisplayName("Get Post Back Returns identity provider results")
+        void getPostCallbackRedirectThrowsWhenValidationFails() {
+            doReturn(mockOAuth).when(oAuthCoordinator).getOAuth();
+            when(mockOAuth.validate(anyString(), anyString(), any(HttpServletResponse.class)))
+                    .thenReturn(false);
+
+            Map<String, String> params = new HashMap<>();
+            params.put("state", "State");
+            params.put("code", "Code");
+
+            assertThrows(UnauthorisedException.class,
+                    () -> oAuthCoordinator.getPostCallbackRedirectURL(mockResponse, params));
+
+            verify(mockOAuth).validate("State", "Code", mockResponse);
```

Comment:
We can probably have constants for both State & Code as they seem to be used across the test class.

Reply:
Good suggestion.